### PR TITLE
feature: support aarch64

### PR DESCRIPTION
**Pull request by luoyunhe**
```release-note
Feature: support aarch64
```

### CODE REVIEW

File: WORKSPACE [hunk 1]
```diff
@@ -455,18 +537,39 @@ http_file(
...
+
 http_file(
     name = "libxcrypt-compat",
     sha256 = "51d74854365a393393b4457e3d92ba103c08671b4c881a8a1d9fcb8a54a4a737",
     urls = ["https://storage.googleapis.com/builddeps/51d74854365a393393b4457e3d92ba103c08671b4c881a8a1d9fcb8a54a4a737"],
 )
 
+http_file(
+    name = "libxcrypt-compat-aarch64",
+    urls = [
```
Comment (awels) [inline WORKSPACE:568]:
```
http_file(
    name = "libxcrypt-compat-aarch64",
    sha256="f36cfd27003bcfc6f5fba9b99414402397f4b57fd503e84514a06d794f313346"
    urls = [
        "http://download.fedoraproject.org/pub/fedora/linux/updates/33/Everything/aarch64/Packages/l/libxcrypt-compat-4.4.23-1.fc33.aarch64.rpm",
    ],
)
```

File: WORKSPACE [hunk 2]
```diff
@@ -503,8 +606,20 @@ http_file(
     urls = ["https://storage.googleapis.com/builddeps/049778a480dd02774934b37c127b345d8748bfbec1e584f9c412a41af34eaf89"],
 )
 
+http_file(
+    name = "liburing-aarch64",
+    urls = ["https://mirrors.tuna.tsinghua.edu.cn/fedora/updates/33/Everything/aarch64/Packages/l/liburing-0.7-3.fc33.aarch64.rpm"],
```
Comment (awels) [inline WORKSPACE:611]:
```
http_file(
    name = "liburing-aarch64",
    sha256="253d0c1dc3180f44766c298ae4cd3426ec7a60a41ea0dc50d0884928b031c1b7"
    urls = [
        "http://download.fedoraproject.org/pub/fedora/linux/updates/33/Everything/aarch64/Packages/l/liburing-0.7-3.fc33.aarch64.rpm",
    ],
)
```

File: WORKSPACE [hunk 3]
```diff
@@ -503,8 +606,20 @@ http_file(
...
+
 http_file(
     name = "libseccomp",
     sha256 = "964e39835b59c76b7eb3f78c460bfc6e7acfb0c40b901775c7e8a7204537f8a7",
     urls = ["https://storage.googleapis.com/builddeps/964e39835b59c76b7eb3f78c460bfc6e7acfb0c40b901775c7e8a7204537f8a7"],
 )
+
+http_file(
+    name = "libnbd-aarch64",
+    urls = [
```
Comment (awels) [inline WORKSPACE:622]:
```
http_file(
    name = "libnbd-aarch64",
    sha256="0e24ffdcda4efc7445834c86d29d7416426bafa9073259cc31bcb1f03fb5eaeb"
    urls = [
        "http://download.fedoraproject.org/pub/fedora/linux/updates/33/Everything/aarch64/Packages/l/libnbd-1.6.5-1.fc33.aarch64.rpm",
    ],
)
```

**Review by awels — COMMENTED** [review]
Hi,

Thanks for the PR, could you update the rpms as I put in the review. In particular point at the fedora download mirrors. I looked up the sha256s for the rpms as well so you should be able to just copy and paste.

Since we have no aarch64 lane at this point, I am looking into how we can remedy this.

**Comment by awels** [comment]
So I think the current remaining failures are because we don't have a .bazelrc file like kubevirt does. Looks like 2 things needs to happen.

```
# The ci.bazelrc is an untracked file that can be generated by hack/dockerized or automated/test.sh scripts as part of CI
try-import ci.bazelrc

# Bazel has a rule of precedence so we can specify / overwrite architecture specific commands if needed
build:x86_64 --platforms=@io_bazel_rules_go//go/toolchain:linux_amd64_cgo
run:x86_64 --platforms=@io_bazel_rules_go//go/toolchain:linux_amd64_cgo
test:x86_64 --platforms=@io_bazel_rules_go//go/toolchain:linux_amd64_cgo

build:aarch64 --platforms=@io_bazel_rules_go//go/toolchain:linux_arm64_cgo
run:aarch64 --platforms=@io_bazel_rules_go//go/toolchain:linux_arm64_cgo
test:aarch64 --platforms=@io_bazel_rules_go//go/toolchain:linux_arm64_cgo --host_javabase=@local_jdk//:jdk

build:crossbuild-aarch64 --incompatible_enable_cc_toolchain_resolution --platforms=//bazel/platforms:aarch64-none-linux-gnu --platforms=@io_bazel_rules_go//go/toolchain:linux_arm64_cgo
run:crossbuild-aarch64  --incompatible_enable_cc_toolchain_resolution --platforms=//bazel/platforms:aarch64-none-linux-gnu --platforms=@io_bazel_rules_go//go/toolchain:linux_arm64_cgo
test:crossbuild-aarch64 --platforms=@io_bazel_rules_go//go/toolchain:linux_arm64_cgo --host_javabase=@local_jdk//:jdk

build --define gotags=selinux
```
and we need to modify automation/test.sh to create ci.bazelrc instead of .bazelrc
```
# Create .bazelrc to use remote cache
cat >ci.bazelrc <<EOF
startup --host_jvm_args=-Dbazel.DigestFunction=sha256
build --remote_local_fallback
build --remote_http_cache=${BAZEL_CACHE}
build --jobs=4
EOF
```

**Comment by luoyunhe** [comment]
> So I think the current remaining failures are because we don't have a .bazelrc file like kubevirt does. Looks like 2 things needs to happen.
> 
> ```
> # The ci.bazelrc is an untracked file that can be generated by hack/dockerized or automated/test.sh scripts as part of CI
> try-import ci.bazelrc
> 
> # Bazel has a rule of precedence so we can specify / overwrite architecture specific commands if needed
> build:x86_64 --platforms=@io_bazel_rules_go//go/toolchain:linux_amd64_cgo
> run:x86_64 --platforms=@io_bazel_rules_go//go/toolchain:linux_amd64_cgo
> test:x86_64 --platforms=@io_bazel_rules_go//go/toolchain:linux_amd64_cgo
> 
> build:aarch64 --platforms=@io_bazel_rules_go//go/toolchain:linux_arm64_cgo
> run:aarch64 --platforms=@io_bazel_rules_go//go/toolchain:linux_arm64_cgo
> test:aarch64 --platforms=@io_bazel_rules_go//go/toolchain:linux_arm64_cgo --host_javabase=@local_jdk//:jdk
> 
> build:crossbuild-aarch64 --incompatible_enable_cc_toolchain_resolution --platforms=//bazel/platforms:aarch64-none-linux-gnu --platforms=@io_bazel_rules_go//go/toolchain:linux_arm64_cgo
> run:crossbuild-aarch64  --incompatible_enable_cc_toolchain_resolution --platforms=//bazel/platforms:aarch64-none-linux-gnu --platforms=@io_bazel_rules_go//go/toolchain:linux_arm64_cgo
> test:crossbuild-aarch64 --platforms=@io_bazel_rules_go//go/toolchain:linux_arm64_cgo --host_javabase=@local_jdk//:jdk
> 
> build --define gotags=selinux
> ```
> 
> and we need to modify automation/test.sh to create ci.bazelrc instead of .bazelrc
> 
> ```
> # Create .bazelrc to use remote cache
> cat >ci.bazelrc <<EOF
> startup --host_jvm_args=-Dbazel.DigestFunction=sha256
> build --remote_local_fallback
> build --remote_http_cache=${BAZEL_CACHE}
> build --jobs=4
> EOF
> ```

thanks.

**Comment by awels** [comment]
Let me review again, but there is one problem remaining after we merge, we have no jobs that build and push aarch64 containers. We will have to create some jobs in github.com/kubevirt/project-infra for containerized-data-import to produce builds. And secondly we should create at least a periodic job with some aarch64 based test runs so we know we didn't break it.

**Comment by zhlhahaha** [comment]
Hi @luoyunhe , nice PR, I have done some tests on ARM64 server, only one unit-test failed.

```
/root/go/src/kubevirt.io/containerized-data-importer/pkg/operator/controller/controller_suite_test.go:46

  Unexpected error:
      <*fmt.wrapError | 0x400079e7c0>: {
          msg: "failed to start the controlplane. retried 5 times: fork/exec /usr/local/kubebuilder/bin/etcd: no such file or directory",
          err: {
              Op: "fork/exec",
              Path: "/usr/local/kubebuilder/bin/etcd",
              Err: 0x2,
          },
      }
      failed to start the controlplane. retried 5 times: fork/exec /usr/local/kubebuilder/bin/etcd: no such file or directory
  occurred

  /root/go/src/kubevirt.io/containerized-data-importer/pkg/operator/controller/controller_suite_test.go:53
------------------------------

Ran 51 of 0 Specs in 0.010 seconds
FAIL! -- 0 Passed | 51 Failed | 0 Pending | 0 Skipped
```
The https://github.com/kubernetes-sigs/kubebuilder/releases does not contain etcd binary for ARM64, which leads to this error.

**Comment by zhlhahaha** [comment]
> Let me review again, but there is one problem remaining after we merge, we have no jobs that build and push aarch64 containers. We will have to create some jobs in github.com/kubevirt/project-infra for containerized-data-import to produce builds. And secondly we should create at least a periodic job with some aarch64 based test runs so we know we didn't break it.

Hi @awels, currently there is only one arm64 server on kubevirt/project-infra, and it is running kubevirt/kubevirt e2e test periodic jobs. I am not sure if there would be some conflicts when we run both kubevirt e2e test and containerized-data-import e2e test at same time in the same machine. (Nested virtualization does not support on ARM64 server, so both tests have to run on bare metal). Otherwise, we can try to give them different time slots, make them run separately.

**Comment by awels** [comment]
One good thing compared to KubeVirt is that CDI does not require nested virtualization. We are not starting any VMs in CDI test suite so we should be able to run a larger suite compared to KubeVirt.

**Comment by awels** [comment]
@zhlhahaha Hi, so I think we can set up a CDI periodic to not run at the same time as KubeVirt. Are you okay with us setting up that periodic job to run on the external server you are providing?

**Comment by zhlhahaha** [comment]
> @zhlhahaha Hi, so I think we can set up a CDI periodic to not run at the same time as KubeVirt. Are you okay with us setting up that periodic job to run on the external server you are providing?

Yes, of course.

**Comment by awels** [comment]
Great, since we don't need virtualization, we should be in pretty good shape after we build some aarch64 images, we can probably skip the functional vddk tests

**Comment by awels** [comment]
Okay, in the meantime, I will merge this PR, so we can get started on building images.
/lgtm
/approve

**Comment by awels** [comment]
We have a nightly build for ARM at this point https://github.com/kubevirt/project-infra/pull/1475 was merged. And I am seeing the images appear on quay.io

**Comment by awels** [comment]
Next will be configuring the test runs.

**Pull request merged**